dev/bots/check_tests_cross_imports.dart

```diff
@@ -0,0 +1,602 @@
+//   bin/cache/dart-sdk/bin/dart --enable-asserts dev/bots/check_tests_cross_imports.dart
+
+import 'dart:io';
+
+import 'package:args/args.dart';
+import 'package:file/file.dart';
+import 'package:file/local.dart';
+import 'package:path/path.dart' as path;
+
+import 'utils.dart';
+
+final String _scriptLocation = path.fromUri(Platform.script);
+final String _flutterRoot = path.dirname(path.dirname(path.dirname(_scriptLocation)));
+final String _testDirectoryPath = path.join(_flutterRoot, 'packages', 'flutter', 'test');
+
+void main(List<String> args) {
+  final ArgParser argParser = ArgParser();
+  argParser.addFlag('help', negatable: false, help: 'Print help for this command.');
+  argParser.addOption(
+    'test',
+    valueHelp: 'path',
+    defaultsTo: _testDirectoryPath,
+    help: 'A location where the tests are found.',
+  );
+  argParser.addOption(
+    'flutter-root',
+    valueHelp: 'path',
+    defaultsTo: _flutterRoot,
+    help: 'The path to the root of the Flutter repo.',
+  );
+  final ArgResults parsedArgs;
+
+  void usage() {
+    print('dart --enable-asserts ${path.basename(_scriptLocation)} [options]');
+    print(argParser.usage);
+  }
+
+  try {
+    parsedArgs = argParser.parse(args);
+  } on FormatException catch (e) {
+    print(e.message);
+    usage();
+    exit(1);
+  }
+
+  if (parsedArgs['help'] as bool) {
+    usage();
+    exit(0);
+  }
+
+  const FileSystem filesystem = LocalFileSystem();
+  final Directory tests = filesystem.directory(parsedArgs['test']! as String);
+  final Directory flutterRoot = filesystem.directory(parsedArgs['flutter-root']! as String);
+
+  final TestsCrossImportChecker checker = TestsCrossImportChecker(
+    testsDirectory: tests,
+    flutterRoot: flutterRoot,
+  );
+
+  if (!checker.checkCodeSamples()) {
+    reportErrorsAndExit('Some errors were found in the API docs code samples.');
+  }
+  reportSuccessAndExit('All examples are linked and have tests.');
+}
+
+class TestsCrossImportChecker {
+  TestsCrossImportChecker({
+    required this.testsDirectory,
+    required this.flutterRoot,
+    this.filesystem = const LocalFileSystem(),
+  });
+
+  final Directory testsDirectory;
+  final Directory flutterRoot;
+  final FileSystem filesystem;
+
+  bool checkCodeSamples() {
+    filesystem.currentDirectory = flutterRoot;
+
+    final Map<Library, Set<File>> filesByLibrary = <Library, Set<File>>{};
+    for (final Library library in Library.values) {
+      filesByLibrary[library] = getTestFiles(testsDirectory, library);
+    }
+
+    final Set<File> widgetsTestsImportingMaterial = getFilesWithImports(
+      filesByLibrary[Library.widgets]!,
+      Library.material,
+    );
+    final Set<File> widgetsTestsImportingCupertino = getFilesWithImports(
+      filesByLibrary[Library.widgets]!,
+      Library.cupertino,
+    );
+    final Set<File> cupertinoTestsImportingMaterial = getFilesWithImports(
+      filesByLibrary[Library.cupertino]!,
+      Library.material,
+    );
+
```

Comment:
What about Material importing Cupertino?

Reply:
Material importing Cupertino is allowed, at least for now. Material will continue to contain platform adaptive code (like Switch.adaptive), and it will also be responsible for testing interactions that involve both Material and Cupertino.